File: Loenn/entities/blueBone.lua
-- Loenn plugin for DZ/BlueBone (chapter 11 - Snowding City).
-- Placeholder visual: blue capsule (rectangle shaft, circle knobs). Swap the sprite
-- function for a drawableSprite once real art exists; no placement data has to change.

local drawableFunction = require("structs.drawable_function")
local drawing = require("utils.drawing")
local utils = require("utils")

local blueBone = {}

blueBone.name = "DZ/BlueBone"
blueBone.depth = -8500
blueBone.nodeLimits = {1, 1}
blueBone.nodeLineRenderType = "line"

blueBone.placements = {
    name = "blue_bone",
    data = {
        boneLength = 48.0,
        boneThickness = 8.0,
        horizontal = false,
        speed = 90.0,
        moveThreshold = 12.0,
        pauseTime = 0.0,
        loop = true,
        flag = "",
        invertFlag = false
    }
}

blueBone.fieldInformation = {
    boneLength = {
        minimumValue = 8.0
    },
    boneThickness = {
        minimumValue = 4.0
    },
    speed = {
        minimumValue = 0.0
    },
    moveThreshold = {
        minimumValue = 0.0
    },
    pauseTime = {
        minimumValue = 0.0
    }
}

local fillColor = {0.23, 0.5, 0.84, 0.9}
local lineColor = {0.75, 0.89, 1.0, 1.0}

-- Matches the C# hitbox: centred on the entity, long axis set by "horizontal".
local function boneRect(entity)
    local length = entity.boneLength or 48
    local thickness = entity.boneThickness or 8

    if entity.horizontal then
        return length, thickness
    end

    return thickness, length
end

local function drawBone(x, y, width, height, thickness)
    drawing.callKeepOriginalColor(function()
        local left, top = x - width / 2, y - height / 2
        local knob = thickness * 0.75

        love.graphics.setColor(fillColor)
        love.graphics.rectangle("fill", left, top, width, height)

        if width > height then
            love.graphics.circle("fill", left, y, knob)
            love.graphics.circle("fill", left + width, y, knob)
        else
            love.graphics.circle("fill", x, top, knob)
            love.graphics.circle("fill", x, top + height, knob)
        end

        love.graphics.setColor(lineColor)
        love.graphics.rectangle("line", left, top, width, height)

        if width > height then
            love.graphics.circle("line", left, y, knob)
            love.graphics.circle("line", left + width, y, knob)
        else
            love.graphics.circle("line", x, top, knob)
            love.graphics.circle("line", x, top + height, knob)
        end
    end)
end

function blueBone.sprite(room, entity)
    local width, height = boneRect(entity)
    local sprites = {
        drawableFunction.fromFunction(drawBone, entity.x or 0, entity.y or 0, width, height,
            entity.boneThickness or 8)
    }

    -- Ghost of the bone at its end point, so the sweep is readable.
    for _, node in ipairs(entity.nodes or {}) do
        table.insert(sprites, drawableFunction.fromFunction(drawBone, node.x, node.y, width, height,
            entity.boneThickness or 8))
    end

    return sprites
end

function blueBone.selection(room, entity)
    local width, height = boneRect(entity)
    local x, y = entity.x or 0, entity.y or 0
    local main = utils.rectangle(x - width / 2, y - height / 2, width, height)
    local nodeRects = {}

    for _, node in ipairs(entity.nodes or {}) do
        table.insert(nodeRects, utils.rectangle(node.x - width / 2, node.y - height / 2, width, height))
    end

    return main, nodeRects
end

return blueBone

File: Source/Entities/Entities/Ch11/BlueBone.cs
namespace Celeste.Entities.Chapters.Ch11
{
    /// <summary>
    /// BlueBone - Undertale blue attack. The bone sweeps between its spawn point and
    /// its node; it only kills the player if the player is <em>moving</em> when it
    /// passes through them. Stand still and it is harmless, so the room becomes a
    /// stop-and-go rhythm puzzle.
    ///
    /// "Moving" means speed above <c>moveThreshold</c>, or dashing at any speed - you
    /// cannot cheese it by dashing into a standstill.
    ///
    /// Placeholder art: blue capsule (square shaft, circle knobs at each end).
    /// </summary>
    [CustomEntity("DZ/BlueBone")]
    [Tracked]
    public class BlueBone : Entity
    {
        #region Fields
        private static readonly Color BoneFill = Calc.HexToColor("3a7fd5");
        private static readonly Color BoneOutline = Calc.HexToColor("bfe2ff");
        private static readonly Color DangerOutline = Calc.HexToColor("ff5a5a");

        private readonly float boneLength;
        private readonly float boneThickness;
        private readonly bool horizontal;
        private readonly float speed;
        private readonly float moveThreshold;
        private readonly float pauseTime;
        private readonly bool loop;
        private readonly string flag;
        private readonly bool invertFlag;

        private readonly Vector2 startPosition;
        private readonly Vector2 endPosition;
        private bool towardEnd = true;
        private float pauseTimer;
        private float pulse;
        private Level level;
        #endregion

        #region Constructor
        public BlueBone(EntityData data, Vector2 offset)
            : base(data.Position + offset)
        {
            boneLength = Math.Max(8f, data.Float("boneLength", 48f));
            boneThickness = Math.Max(4f, data.Float("boneThickness", 8f));
            horizontal = data.Bool("horizontal", false);
            speed = Math.Max(0f, data.Float("speed", 90f));
            moveThreshold = Math.Max(0f, data.Float("moveThreshold", 12f));
            pauseTime = Math.Max(0f, data.Float("pauseTime", 0f));
            loop = data.Bool("loop", true);
            flag = data.Attr("flag", "");
            invertFlag = data.Bool("invertFlag", false);

            startPosition = Position;
            endPosition = data.Nodes != null && data.Nodes.Length > 0
                ? data.Nodes[0] + offset
                : Position;

            Collider = horizontal
                ? new Hitbox(boneLength, boneThickness, -boneLength / 2f, -boneThickness / 2f)
                : new Hitbox(boneThickness, boneLength, -boneThickness / 2f, -boneLength / 2f);

            Depth = -8500;
        }
        #endregion

        #region Public API
        /// <summary>True when the player is moving fast enough for this bone to be lethal.</summary>
        public bool IsLethalTo(Player player)
        {
            if (player == null || player.Dead)
            {
                return false;
            }

            return player.DashAttacking || player.Speed.Length() > moveThreshold;
        }
        #endregion

        #region Private
        /// <summary>Whether this entity's enable flag currently permits it to run.</summary>
        private bool FlagAllows()
        {
            if (string.IsNullOrEmpty(flag) || level == null)
            {
                return true;
            }

            return level.Session.GetFlag(flag) != invertFlag;
        }

        private void Travel()
        {
            if (speed <= 0f || startPosition == endPosition)
            {
                return;
            }

            if (pauseTimer > 0f)
            {
                pauseTimer -= Engine.DeltaTime;
                return;
            }

            Vector2 target = towardEnd ? endPosition : startPosition;
            Vector2 toTarget = target - Position;
            float step = speed * Engine.DeltaTime;

            if (toTarget.Length() <= step)
            {
                Position = target;

                if (!loop && towardEnd)
                {
                    Audio.Play(ChapterSfx.SnowBoneBreak, Center);
                    RemoveSelf();
                    return;
                }

                towardEnd = !towardEnd;
                pauseTimer = pauseTime;
                return;
            }

            Position += toTarget.SafeNormalize() * step;
        }
        #endregion

        #region Entity Overrides
        public override void Added(Scene scene)
        {
            base.Added(scene);
            level = scene as Level;
        }

        public override void Update()
        {
            base.Update();

            if (!FlagAllows())
            {
                pulse = 0f;
                return;
            }

            Travel();

            Player player = Scene.Tracker.GetEntity<Player>();
            if (player != null && CollideCheck(player))
            {
                if (IsLethalTo(player))
                {
                    Audio.Play(ChapterSfx.SnowHit, Center);
                    player.Die((player.Center - Center).SafeNormalize());
                }
                else
                {
                    // Passing safely - flash so the player learns the rule.
                    pulse = 1f;
                }
            }
            else if (player != null && Vector2.Distance(player.Center, Center) < boneLength && IsLethalTo(player))
            {
                pulse = Math.Max(pulse, 0.6f);
            }

            pulse = Calc.Approach(pulse, 0f, Engine.DeltaTime * 3f);
        }

        public override void Render()
        {
            base.Render();

            if (!FlagAllows())
            {
                return;
            }

            Rectangle bounds = Collider.Bounds;
            Color outline = Color.Lerp(BoneOutline, DangerOutline, pulse);

            Placeholder.Square(bounds.X, bounds.Y, bounds.Width, bounds.Height, BoneFill, outline);

            // Bone knobs at both ends.
            float knob = boneThickness * 0.75f;
            if (horizontal)
            {
                Placeholder.Circle(new Vector2(bounds.Left, Center.Y), knob, BoneFill, outline);
                Placeholder.Circle(new Vector2(bounds.Right, Center.Y), knob, BoneFill, outline);
            }
            else
            {
                Placeholder.Circle(new Vector2(Center.X, bounds.Top), knob, BoneFill, outline);
                Placeholder.Circle(new Vector2(Center.X, bounds.Bottom), knob, BoneFill, outline);
            }
        }

        public override void DebugRender(Camera camera)
        {
            base.DebugRender(camera);
            Draw.Line(startPosition, endPosition, Color.Aqua * 0.5f);
        }
        #endregion
    }
}
